src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -141,21 +141,12 @@ def resolve(
                     and candidate.source_link.ext != ".zip"
                 )
                 if looks_like_sdist:
-                    # is a local sdist -- show a deprecation warning!
-                    reason = (
-                        "Source distribution is being reinstalled despite an "
-                        "installed package having the same name and version as "
-                        "the installed package."
-                    )
-                    replacement = "use --force-reinstall"
-                    deprecated(
-                        reason=reason,
-                        replacement=replacement,
-                        gone_in="21.3",
-                        issue=8711,
+                    # is a local sdist -- disallow.
+                    raise SamePackageAlreadyInstalled(
+                        refused_to_install=candidate.source_link.file_path
```

Comment:
Should that really be an error ? Or just a `Requirement already satisfied: ...` message ?